### PR TITLE
Alexej/eng 35 run examples from cli

### PR DESCRIPTION
**Pull request by AlexejPenner**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Initial PR with logic, unittests will follow

### CODE REVIEW

File: src/zenml/cli/example.py [hunk 1]
```diff
@@ -176,9 +294,45 @@ def examples(self) -> List[Example]:
                 not name.startswith(".")
                 and not name.startswith("__")
                 and not name.startswith("README")
+                and not name.endswith(".sh")
             )
         ]
 
+    @property
+    def is_installed(self) -> bool:
+        """Checks if the"""
```
Comment (htahir1) [inline src/zenml/cli/example.py:303]: Missing docstring :-)

**Review by htahir1 — CHANGES_REQUESTED** [review]
Excellent PR! I have some minor comments and changes, but overall the bash foo is awesome!

File: src/zenml/cli/example.py [hunk 2]
```diff
@@ -41,42 +42,152 @@
...
+        location
+
+        Args:
+            bash_file: File location of the bash script to run examples
+            force: Whether to force the install
+        """
+        if fileio.file_exists(bash_file):
+            os.chdir(self.path)
+            try:
+                # TODO[HIGH] Catch errors that might be thrown in subprocess
```
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/cli/example.py:114]: also needs the `:` after the brackets

File: src/zenml/cli/example.py [hunk 3]
```diff
@@ -306,56 +469,106 @@ def info(git_examples_handler: GitExamplesHandler, example_name: str) -> None:
...
+    git_examples_handler: GitExamplesHandler,
+    example_name: str,
+    path: str,
+    force: bool,
+) -> None:
+    """Run the example at the specified relative path.
+    `zenml example pull EXAMPLE_NAME` has to be called with the same relative
+    path before the run command.
+    """
+    # TODO[MEDIUM] - create a post_run function inside individual setup.sh
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/cli/example.py:553]: will fix

File: src/zenml/cli/example.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/cli/example.py:114]: will fix

File: src/zenml/cli/example.py [hunk 4]
```diff
@@ -41,42 +42,152 @@
...
+                        cwd=self.path,
+                    )
+            except RuntimeError:
+                raise NotImplementedError(
+                    f"Currently the example {self.name} "
+                    "has no implementation for the "
+                    "run method"
+                )
+            except subprocess.CalledProcessError as e:
+                if str(e.returncode) == "42":
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/cli/example.py:142]: We are throwing this specific error code here: https://github.com/zenml-io/zenml/blob/9fd2cbc0c3a0f4f9379a2993cba28c9b7a536b9d/examples/standard_interfaces/setup.sh#L6

We don't want to simply catch all errors here, so I came up with an errorcode we can use in the bash scripts to throw NotImplemented errors. Will change it to 38 - as this seems to be the Function Not Implemented Errorcode used in some other libraries.

Will be changing it to an int-comparison, seems like the e.returncode is an integer by default

File: src/zenml/cli/example.py [hunk 5]
```diff
@@ -306,56 +469,106 @@ def info(git_examples_handler: GitExamplesHandler, example_name: str) -> None:
...
+    path before the run command.
+    """
+    # TODO[MEDIUM] - create a post_run function inside individual setup.sh
+    #  to inform user how to clean up
+    examples_dir = os.path.join(os.getcwd(), path)
+    try:
+        _ = git_examples_handler.get_examples(example_name)[0]
+    except KeyError as e:
+        error(str(e))
+    else:
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/cli/example.py:560]: Yes intentionally using it. I feel it makes the flow control a little bit more explicit. Else only gets executed if the exception does not occur. This is explicitly opposite behavior to finally.

 Further above in the info function i do the following
```python
    try:
        **example_obj** = git_examples_handler.get_examples(example_name)[0]

    except KeyError as e:
        error(str(e))

    else:
        title(example_obj.name)
        pretty_print(**example_obj**.readme_content)
```

Here the else has the additional small advantage, that its immediately obvious that the example_obj from within the try clause is accessed from within the else clause. Pycharm would complain here if the else clause wasn't present.

If you don't like it we can remove it though.

File: src/zenml/cli/example.py [hunk 6]
```diff
@@ -41,42 +42,152 @@
...
+        return len(self.python_files_in_dir) > 0
+
+    @property
+    def executable_python_example(self) -> str:
+        """Return the python file for the example"""
+        if self.has_single_python_file:
+            return self.python_files_in_dir[0]
+        elif self.has_any_python_file:
+            raise RuntimeError(
+                "Unclear which python file to return for "
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/cli/example.py:92]: My approach here was, to have the explicit error message that directly states the issue at the deeper levels of the code, at the cli level the error message is caught and translated into an error message that tells the user this:

```python
            except RuntimeError:
                raise NotImplementedError(
                    f"Currently the example {self.name} "
                    "has no implementation for the "
                    "run method"
                )
```

With this approach i want to make easily debuggable for anyone looking to fix this without confusing the user that just wants to run an example.

File: src/zenml/cli/example.py (hunk 1 above)
Comment (AlexejPenner) [reply src/zenml/cli/example.py:303]: Ah I missed this one, the method is no longer used. Will remove it altogether

**Review by htahir1 — APPROVED** [review]
LGTM!

**Pull request merged**